### DISCUSSION TITLE
embedded in SpringBoot does not find files **/*.log, only spring.log

### DISCUSSION THREAD

**Discussion by EdithBirrer1** [sevdokimov/log-viewer#87] [Q&A]
I have embedded log-viewer into SpringBoot application. 

We have a logback- configuration which renames log files every night, so there are logfiles like
- `spring.log.2021-11-29.log`
- `spring.log.2021-11-30.log`
- `spring.log.2021-11-01.log`
- `spring.log`  (this is the one of "today")

The documentation says: 
"The default configuration gives access to all files with ".log" extension: **/*.log"
However, in the frontend I can only see `spring.log`, not the others. If I write a  `LogConfigurationLoader` (as in your example "demo/spring-boot-2.0") and list the files of the log directory, I get the other files listed as "favourites". But this list is calculated only at server startup, so I will never show files created afterwards. 

How can I get access to all files named *.log as the documentation says? (The documentation describes how to write a configuration file `log-viewer-1.0.1/config.conf`, but how to do it when embedded in SpringBoot application?) 

The app looks great! So I would like to use it, but I need all files. Thanks a lot!

**Comment by sevdokimov** [comment]
The purpose of `LogConfigurationLoader` is to load configured logs. Actually, you have one configured log only: "spring.log". "spring.log.2021-11-29.log" is a part of "spring.log". Log-Viewer must load "spring.log.*.log" automatically when you are open "spring.log". But this functionality is not implemented yet.

As a workaround, you can create a spring bean implementing `com.logviewer.api.LvPathResolver`, this bean will be used to resolve the path from URL to a real file on the disk. You will open "http:/ /localhost:8090/logs/log?path=my-logs" and "my-logs" will be resolved to the list of file located in "my-app/logs" directory (spring.log, spring.log.2021-11-29.log, spring.log.2021-11-30.log, ...)

```
public class MyPathResolver implements LvPathResolver {

    private final File logDirectory;

    public MyPathResolver(File logDirectory) {
        this.logDirectory = logDirectory;
    }

    @Override
    public Collection<LogPath> resolvePath(String pathFromHttpParameter) {
        if (pathFromHttpParameter.equals("my-logs")) {
            return Stream.of(logDirectory.listFiles())
                    .filter(f -> f.getName().endsWith(".log") && f.isFile())
                    .map(f -> new LogPath(null, f.getAbsolutePath()))
                    .collect(Collectors.toList());
        }
        
        return null;
    }
}
```

Also, you have to override the security manager, by default it hides all files except the configured logs ("spring.log" only is visible). You have to override the bean created in `com.logviewer.config.LogViewerAutoConfig.lvLogManager()` , or call on it the following code:
`res.setPaths(Collections.singletonList(PathPattern.directory(myLogDirectory)))`.

I must to improve the documentation...

**Reply by EdithBirrer1** [reply]
Thank you! In fact, I it was not even necessary to use a path resolver. Overriding the SecurityManager did the trick, then I see all the files and can add them.  
I added in my Application.java:
```
private final Path logDirectory = Paths.get("logs").toAbsolutePath();
[...]
/**
 * Adds the files in logDirectory to the list of logs visible by log-viewer.
 * Annotation @Primary is needed because this should overwrite the bean already defined in
 * io\github\sevdokimov\logviewer\log-viewer\1.0.1\log-viewer-1.0.1.jar!\com\logviewer\config\LogViewerAutoConfig.class
 */
@Bean @Primary
public LvFileAccessManagerImpl lvLogManager2() {
    LvFileAccessManagerImpl res = new LvFileAccessManagerImpl(null);
    res.setPaths(Collections.singletonList(PathPattern.directory(logDirectory)));
    return res;
}
```
Of course it would be cool if all "spring.log.*.log" would be automatically combined and displayed when "spring.log" is opened.
And yes, some more documentation or embedding examples would be greatly appreciated! But you made up for it by answering quickly, thanks a lot!

**Comment by sevdokimov** [comment]
> How can I get access to all files named *.log as the documentation says? (The documentation describes how to write a configuration file log-viewer-1.0.1/config.conf, but how to do it when embedded in SpringBoot application?)

Log viewer loads configuration properties from the spring context, you can define the properties as regular spring properties.
